pkg/controller/controller.go

```diff
@@ -1351,9 +1349,68 @@ func (s *Service) ListSnapshots(ctx context.Context, req *csi.ListSnapshotsReque
 	}, nil
 }
 
+func pauseMetroSession(ctx context.Context, metroSessionID string, arr *array.PowerStoreArray) (paused bool, err error) {
+	metroSession, err := arr.Client.GetReplicationSessionByID(ctx, metroSessionID)
+	if err != nil {
+		return false, fmt.Errorf("could not get metro replication session %s", metroSessionID)
+	}
+
+	// confirm the session is in a state we can pause from.
+	if metroSession.State != gopowerstore.RsStateOk &&
+		metroSession.State != gopowerstore.RsStateSynchronizing &&
+		metroSession.State != gopowerstore.RsStatePaused &&
+		metroSession.State != gopowerstore.RsStateSystemPaused &&
+		metroSession.State != gopowerstore.RsStateFractured {
+		return false, fmt.Errorf("could not pause the metro replication session, %s, because the session is not in expected state to pause", metroSession.ID)
+	}
+
+	if metroSession.State != gopowerstore.RsStatePaused {
+		log.Debugf("pausing metro replication session, %s", metroSession.ID)
+
+		// pause the replication session
+		_, err := arr.Client.ExecuteActionOnReplicationSession(ctx, metroSession.ID, gopowerstore.RsActionPause, nil)
+		if err != nil {
+			return false, fmt.Errorf("metro replication session, %s, could not be paused: %s", metroSession.ID, err.Error())
+		}
+	} else {
+		log.Debugf("metro replication session, %s, already paused", metroSession.ID)
+	}
+	return true, nil
+}
+
+func resumeMetroSession(ctx context.Context, metroSessionID string, array *array.PowerStoreArray) (resumed bool, err error) {
+	metroSession, err := array.Client.GetReplicationSessionByID(ctx, metroSessionID)
+	if err != nil {
+		return false, fmt.Errorf("could not get metro replication session: %s", err.Error())
+	}
+
+	// nothing to do if not paused
+	if metroSession.State == gopowerstore.RsStateOk ||
+		metroSession.State == gopowerstore.RsStateSynchronizing ||
+		metroSession.State == gopowerstore.RsStateResuming ||
+		metroSession.State == gopowerstore.RsStateSwitchingToMetroSync ||
+		metroSession.State == gopowerstore.RsStateFractured {
+		log.Debugf("metro replication session, %s, already resumed", metroSession.ID)
+		return false, nil
+	}
+
+	// metro session can only be resumed if it is in 'paused' state
+	if metroSession.State != gopowerstore.RsStatePaused {
+		return false, errors.New("the metro session must be in 'paused' state before resuming")
```

Comment:
Is this the way this is done in other places in the driver? If the desired state is the current state then error? Does the caller check the boolean and do something different?

Reply:
The session could be in states other than 'pause' where the API may fail. We are just checking that and failing before-hand. 
That boolean is just for its caller to return appropriate success response back to K8s.